RecoEgamma/EgammaElectronAlgos/src/ConversionFinder.cc

```diff
@@ -0,0 +1,298 @@
+    float y = .5 * (ya1 + ya2);
+    float rconv = sqrt(pow(x, 2) + pow(y, 2));
+    // The z-position of the conversion is unused, but here is how it could be computed if needed:
+    // float z = ele.dz() + rEl * ele.pz() * std::acos(1 - pow(rconv, 2) / (2. * pow(rEl, 2))) / ele.pt();
+
+    //now assign a sign to the radius of conversion
+    float tempsign = ele.px() * x + ele.py() * y;
+    tempsign = tempsign / std::abs(tempsign);
+    rconv = tempsign * rconv;
+
+    //return an instance of ConversionInfo, but with a NULL track refs
+    return {dist, dcot, rconv, std::nullopt, std::nullopt, -9999, -9999};
+  }
+
+  //------------------------------------------------------------------------------------
+
+  //takes in a vector of candidate conversion partners
+  //and arbitrates between them returning the one with the
+  //smallest R=sqrt(dist*dist + dcot*dcot)
+  ConversionInfo const& arbitrateConversionPartnersbyR(const std::vector<ConversionInfo>& convCandidates) {
+    ConversionInfo const* closestConversion = &convCandidates.front();
+
+    if (convCandidates.size() == 1)
+      return *closestConversion;
+
+    float R = pow(closestConversion->dist, 2) + pow(closestConversion->dcot, 2);
+
+    for (auto const& temp : convCandidates) {
+      float temp_R = pow(temp.dist, 2) + pow(temp.dcot, 2);
+      if (temp_R < R) {
+        R = temp_R;
+        closestConversion = &temp;
+      }
+    }
+
+    return *closestConversion;
+  }
+
+  //------------------------------------------------------------------------------------
+  ConversionInfo findBestConversionMatch(const std::vector<ConversionInfo>& v_convCandidates) {
+    using namespace std;
+
+    if (v_convCandidates.empty())
+      return ConversionInfo{-9999., -9999., -9999., std::nullopt, std::nullopt, -9999, -9999};
+
+    if (v_convCandidates.size() == 1)
+      return v_convCandidates.at(0);
+
+    vector<ConversionInfo> v_0;
+    vector<ConversionInfo> v_1;
+    vector<ConversionInfo> v_2;
+    vector<ConversionInfo> v_3;
+    //loop over the candidates
+    for (unsigned int i = 1; i < v_convCandidates.size(); i++) {
+      ConversionInfo temp = v_convCandidates.at(i);
+
+      if (temp.flag == 0) {
+        bool isConv = false;
+        if (std::abs(temp.dist) < 0.02 && std::abs(temp.dcot) < 0.02 && temp.deltaMissingHits < 3 &&
+            temp.radiusOfConversion > -2)
+          isConv = true;
+        if (sqrt(pow(temp.dist, 2) + pow(temp.dcot, 2)) < 0.05 && temp.deltaMissingHits < 2 &&
+            temp.radiusOfConversion > -2)
+          isConv = true;
+
+        if (isConv)
+          v_0.push_back(temp);
+      }
+
+      if (temp.flag == 1) {
+        if (sqrt(pow(temp.dist, 2) + pow(temp.dcot, 2)) < 0.05 && temp.deltaMissingHits < 2 &&
+            temp.radiusOfConversion > -2)
+          v_1.push_back(temp);
+      }
+      if (temp.flag == 2) {
+        if (sqrt(pow(temp.dist, 2) + pow(temp.dcot * temp.dcot, 2)) < 0.05 && temp.deltaMissingHits < 2 &&
```

Comment:
This `dcot` to the power of four here was also in the original code and might be a mistake. Since this PR aims for optimization without changing the reco output, I didn't do anything about it. Just wanted to raise awareness.

Reply:
I would say it is mistake that must be eventually fixed!
I don't know who's the original author of the code, but at least egamma should have a look: @afiqaize @SohamBhattacharya